File: ui/components/ui/ProducerSession/producer-session/producer-session.ts
import * as pb from '../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { ProducerSessionTask, producerSessionTaskFromPb, producerSessionTaskToPb } from './producer-session-task/producer-session-task';

export type ProducerSession = {
  type: 'producer-session',
  sessionId: string,
  tasks: ProducerSessionTask[]
};

export function producerSessionFromPb(v: pb.ProducerSession): ProducerSession {
  return {
    type: 'producer-session',
    sessionId: v.getSessionId(),
    tasks: v.getTasksList().map(producerSessionTaskFromPb)
  };
}

export function producerSessionToPb(v: ProducerSession): pb.ProducerSession {
  const producerSessionPb = new pb.ProducerSession();
  producerSessionPb.setSessionId(v.sessionId);
  producerSessionPb.setTasksList(v.tasks.map(producerSessionTaskToPb));

  return producerSessionPb;
}

File: server/src/main/scala/producer/producer_session/ProducerSession.scala
package producer.producer_session

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import _root_.producer.producer_session.producer_session_task.ProducerSessionTask

case class ProducerSession(
    sessionId: String,
    tasks: Vector[ProducerSessionTask]
)

object ProducerSession:
    def fromPb(v: pb.ProducerSession): ProducerSession =
        ProducerSession(
            sessionId = v.sessionId,
            tasks = v.tasks.map(ProducerSessionTask.fromPb).toVector
        )

    def toPb(v: ProducerSession): pb.ProducerSession =
        pb.ProducerSession(
            sessionId = v.sessionId,
            tasks = v.tasks.map(ProducerSessionTask.toPb)
        )
